rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,283 @@
+[summary]: #summary
+
+Inspired by the various derivation checks like for broken and insecure packages, a new system called "problems" is introduced. It is planned to eventually replace the previously mentioned systems where possible, as well as the current – undocumented – "warnings" (which currently only prints a trace message for unmaintained packages). A `config.problemHandler` option is added to the nixpkgs configuration, with centralized and granular control over how to handle problems that arise: "throw" (fail evaluation), "trace" (print a warning message) or "ignore" (do nothing).
+
+Additionally, `meta.problems` is added to derivations, which can be used to manually declare that a package has a certain problem. This will then be used to inform users about packages that are in need of maintenance, for example security vulnerabilities or deprecated dependencies.
+
+Using the newly introduced features, we may create a process for removing packages from nixpkgs that is easier to maintain and friendlier to users than just replacing packages with `throw`.
+
+## Motivation
+[motivation]: #motivation
+
+Nixpkgs has the problem that it is often treated as "append-only", i.e. packages only get added but not removed. There are a lot of packages that are broken for a long time, have end-of-life dependencies with known security vulnerabilities or that are otherwise unmaintained.
+
+Let's take the end of life of Python 2 as an example. (This applies to other ecosystems as well, and will come up again and again in the future.) It has sparked a few bulk package removal actions by dedicated persons, but those are pretty work intensive and prone to burn out maintainers. A goal of this RFC is to provide a way to notify all users of a package about the outstanding issues. This will hopefully draw more attention to abandoned packages, and spread the work load. It can also help soften the removal of packages by providing a period for users to migrate away at their own pace.
+
+For some use cases, like for packages without maintainers, we do not want to break evaluation of a package and simply warn the user instead. We want the users to configure all these according to their needs and through a single standardized interface.
+
+## Detailed design
+[design]: #detailed-design
+
+### Package problems
+
+A new attribute is added to the `meta` section of a package: `problems`. If present, it is a list of attrsets which each have at least the following fields:
+
+- `kind`: Required. If present, the resulting warning will be printed as `kind: message`.
+- `message`: Required. A string message describing the issue with the package. The value should:
+  - Start with the "This package", "The application" or equivalent, or simply with the package name.
+  - Be capitalized (unless it starts with the package name).
+  - Use a period at the end.
+- `name`: Required if there are multiple values of the same `kind`. Give the issue a custom name for more easy filtering
```

Comment:
Coming back to [my suggestion of using an attribute set](https://github.com/NixOS/rfcs/pull/127#discussion_r995817447), how about using the attribute name as the kind if it's valid? E.g.

```nix
meta.problems.removed = {
    message = "The application has been abandoned upstream, use libfoo instead";
    date = "1970-01-01";
}
```

would automatically be kind `removed`. Additionally specifying a conflicting kind would give an error.

This way you can always guarantee unique names, no need to use a list, and there's no extra verbosity.

Reply:
Some similar ideas are discussed in the Alternatives section: https://github.com/NixOS/rfcs/pull/127/files/1e6725ece081e4828883a418a1f9a1afbd4dcf55#diff-e089f932103f6decfd159bd39420f43fd834106b788fa06956e771ffd0cda991R219-R248

With your proposal, how would you handle problem kinds where there can be multiple instances, like for security issues?

Reply:
Then you explicitly specify the kind:

```nix
{
  problems.foo = {
    kind = "insecure";
    message = "...";
    date = "...";
  };
  problems.bar = {
    kind = "insecure";
    message = "...";
    date = "...";
  };

  # Could also still have
  problems.insecure = {
    message = "...";
    date = "...";
  };
}
```

Reply:
Thinking more about this, for backwards/forwards compatibility no error or warning should be thrown when both the name is a valid kind and `kind =` is specified. In module system terms, it should work like this:

```nix
{ lib, ... }: {
  options.problems = lib.mkOption {
    default = {};
    type = lib.types.attrsOf (lib.types.submodule ({ name, options, ... }: {

      options.kind = lib.mkOption {
        type = lib.types.enum [
          "removal"
          "deprecated"
          "maintainerless"
          "insecure"
          "broken"
          "unsupported"
        ];
      };

      config.kind = lib.mkIf
        (options.kind.type.check name)
        (lib.mkDefault name);

    }));
  };
}
```

Reply:
I think this approach is less intuitive / more complicated to teach than the currently proposed, but it also has its merits.

Reply:
And I'm just realizing that this is related to [my insistence on not supporting wildcards](https://github.com/NixOS/rfcs/pull/127#discussion_r1056721259), because problem handlers being able to ignore an entire kind of problems is the same as a wildcard and it comes with the same problems (implicitly handling new problems of the same kind that were never shown to the user).

Reply:
I changed the description of `meta.problems` according to your proposal. Let's keep the wildcards discussion to the other thread.